### DISCUSSION TITLE
How to intentionally fail transaction using Foundry?

### DISCUSSION THREAD

**Discussion by Niferu** [Cyfrin/foundry-full-course-cu#257] [Q&A] [answered]
Hello,

I was trying to test this already using brownie and hardhat frameworks and there is a way to do it but only by modifying contract code and adding also another contract, which is not what we want obviously. This is some more advanced stuff and I was wondering if we can do it accurately using foundry?

Let's say we have simple withdraw function:
```solidity
    function withdraw() public onlyOwner {
        for (uint256 funderIndex = 0; funderIndex < s_funders.length; funderIndex++) {
            address funder = s_funders[funderIndex];
            s_addressToAmountFunded[funder] = 0;
        }
        s_funders = new address[](0);

        (bool success, ) = i_owner.call{value: address(this).balance}("");
        require(success);
    }
```

How can we test case where `success` is failing? I assume we have to intentionally fail this transaction but how can we do so using foundry without modifying existing contract and without adding any other contract? Is it even possible?

**Comment by shikhar229169** [comment]
We can make ```success``` to false if we try to send money to a contract, that don't have receive and fallback functions. So making the Test contract the owner, will make success to false.
What you can do is, add a function that allows the owner of contract to transfer ownership, so you can transfer the ownership to Test contract and then success will become false, if you call the withdraw() function.

**Reply by Niferu** [reply]
@shikhar229169 

I know about that as I mentioned in my post I have already tested contracts like that before using different frameworks. The point is if we can test it without modifying our contract or creating other contract without receive/fallback as this is just pointless. I was hoping that foundry finally have such solution maybe, so we can use it like `vm.expectRevert();` before desired transaction to intentionally fail it.

**Comment by Niferu** [comment, marked as answer]
I have managed to do something similar below, but my coverage is still pointing that I have missing branch...

```solidity
    function testSuccessFail() public funded {
        vm.startPrank(fundMe.getOwner());
        vm.mockCallRevert(address(fundMe), abi.encodeWithSignature("withdraw()"), "Transfer Failed!");
        vm.expectRevert("Transfer Failed!");
        fundMe.withdraw();
        vm.stopPrank();
    }
```

**Comment by Niferu** [comment]
I was working on StableCoin tests and that gave me an idea how can we actually test this function properly, so `test coverage` will finally show 100%!

First create `MockFailContract` as below (no fallback/receive):
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

contract MockFailedCall {}
```

Then our test will look like below:
```solidity
    function testRevertsIfTransferFail() public {
        MockFailedCall mock = new MockFailedCall();

        deal(address(mock), 10 ether);

        vm.startPrank(address(mock));
        FundMe mockFundMe = new FundMe(priceFeed);

        mockFundMe.fund{value: SEND_VALUE}();

        vm.expectRevert("Transfer Failed!");
        mockFundMe.withdraw();

        vm.expectRevert(FundMe.FundMe__TransferFailed.selector);
        mockFundMe.withdrawTwo();

        vm.expectRevert();
        mockFundMe.cheaperWithdraw();

        vm.stopPrank();
    }
```

Where all 3 withdraw functions looks like that:
```solidity
    function withdraw() public onlyOwner {
        for (uint256 funderIndex = 0; funderIndex < s_funders.length; funderIndex++) {
            address funder = s_funders[funderIndex];
            s_addressToAmountFunded[funder] = 0;
        }
        s_funders = new address[](0);
        // Transfer vs call vs Send
        // payable(msg.sender).transfer(address(this).balance);
        (bool success, ) = i_owner.call{value: address(this).balance}("");
        require(success, "Transfer Failed!");
    }

    function withdrawTwo() public onlyOwner {
        for (uint256 funderIndex = 0; funderIndex < s_funders.length; funderIndex++) {
            address funder = s_funders[funderIndex];
            s_addressToAmountFunded[funder] = 0;
        }
        s_funders = new address[](0);
        // Transfer vs call vs Send
        // payable(msg.sender).transfer(address(this).balance);
        (bool success, ) = i_owner.call{value: address(this).balance}("");
        if (!success) revert FundMe__TransferFailed();
    }

    function cheaperWithdraw() public onlyOwner {
        address[] memory funders = s_funders;
        // mappings can't be in memory, sorry!
        for (uint256 funderIndex = 0; funderIndex < funders.length; funderIndex++) {
            address funder = funders[funderIndex];
            s_addressToAmountFunded[funder] = 0;
        }
        s_funders = new address[](0);
        // payable(msg.sender).transfer(address(this).balance);
        (bool success, ) = i_owner.call{value: address(this).balance}("");
        require(success);
    }
```


Results:
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/110537615/0985c371-3436-4c7a-ba5e-3a262a872629)

The only line not tested is here: 
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/110537615/222d5fb0-e96a-47ea-a41d-29a799ae8eca)

Does anyone have idea why is it not tested like `require`?

**Reply by theirrationalone** [reply]
sometimes stringified errors as require does `require(_condition, "some error")` may cause difficulty in testing.... that's why we use custom errors who in return have selectors also.

**Reply by Niferu** [reply]
Yes but as you can see I have successfully tested `require` but I'm having problem to test custom error as it doesnt see `!success` and see only custom error that occured, not sure if we can hit that !success to be visible for test coverage

**Reply by theirrationalone** [reply]
All goes to powerful protocol implementation, and overpowered custom errors.... The only main reason why patrick haven't added `fallback` & `receive`and custom error into account to demonstrate beyond the boundry. 

If he adds em'all, then he will fail to demonstrate limited knowledge up-to the Lesson and as an impact, interactions withdraw script will go failed. 

The idea you're using is kind of making a new user using `makeAddr`, whoa woowww, you got some deep foundry knowledge, Actually you have created a cheat code `MockFailedCall`

So the conclusion is, we should follow @shikhar229169 's words Or let tests remain to 87+ < 100.

Wowww, How Great discussion is this, Salute to your scratchings under the Ground. ❤

**Reply by Niferu** [reply]
Thanks a lot @theirrationalone!

I like to understand everything I'm doing haha thats why I'm digging :D

**Reply by cebidhem** [reply]
Thanks @Neftyr for the hint on testing reverts. I wrapped my head around this without coming close to a solution, I could only test `FundMe_NotOwner`.

Quick thing though on the code you've shared: the function `testRevertsIfTransferFail()` doesn't know `priceFeed` (except if you built it in another part), so I added something like that on my end:
```solidity
        HelperConfig helperConfig = new HelperConfig();
        address ethUsdPriceFeed = helperConfig.activeNetworkConfig();

        deal(address(mockFailedCall), 10 ether);

        vm.startPrank(address(mockFailedCall));
        // set address(mock) as owner
        FundMe mockFundMe = new FundMe(ethUsdPriceFeed);
```

with the import of `HelperConfig`
```solidity
import {HelperConfig} from "../../script/HelperConfig.s.sol";
```

But again, thanks a lot for the hint as I couldn't find a good way to mock a revert here after trying the cheatcode `mockCallRevert`.

@PatrickAlphaC I think this [comment](https://github.com/Cyfrin/foundry-full-course-f23/discussions/257#discussioncomment-6831609) should be marked as the answer as it pushes branch coverage to 100%. wdyt ?

**Reply by Niferu** [reply]
Hi @cebidhem 

Yes I have included `priceFeed` in another part, I have just shared example look of my functions that I was trying to test ;)

It was challenge for me for sooo long too! Thru 3 courses python version, java script version and now foundry I was trying to get this test done. Finally foundry gave me tools to solve this test! I'm so glad I'm not alone doing that stuff and I'm so happy I could help you cheers mate!

**Comment by theirrationalone** [comment]
I tried to hack patrick's this implementation so many times.... But here @shikhar229169 is right. Even if we try to modify patrick's implementation then it would not be considered a hack therefore we changed the protocol completely by our intentional vulnerability. 

That's why PULL over Push is considered. 

Not to pay but withdraw instead. 

Thanks for the cheat code btw.
